lib/videojs-playlist-ui.js

```diff
@@ -1,25 +1,188 @@
+      for (let prop in variant) {
+        source[prop] = variant[prop];
+      }
+      picture.appendChild(source);
+    }
+
+    // the default version of a <picture> is specified by an <img>
+    let variant = thumbnail[thumbnail.length - 1];
+    let img = document.createElement('img');
+    for (let prop in variant) {
+      img[prop] = variant[prop];
+    }
+    picture.appendChild(img);
+  }
+  return picture;
+};
+
+videojs.PlaylistMenuItem = videojs.Component.extend({
+  init: function(player, options) {
+    if (!options.item) {
+      throw new Error('Cannot construct a PlaylistMenuItem without an item option');
+    }
+    // stub out the element so Component doesn't construct one
+    options.el = true;
+    videojs.Component.call(this, player, options);
+    this.el_ = this.createEl(options.item);
+
+    this.item_ = options.item;
+    this.emitTapEvents();
+
+    this.on(['click', 'tap'], (event) => {
+      const currentIndex = player.playlist().indexOf(this.item_);
+      if (currentIndex >= 0) {
+        player.playlist.currentItem(currentIndex);
+      }
+    });
+  },
+  createEl: function(item) {
+    let li = document.createElement('li');
+    li.className = 'vjs-playlist-item';
+
+    // Thumbnail image
+    li.appendChild(createThumbnail(item.thumbnail));
+
+    // Duration
+    if (item.duration) {
+      let duration = document.createElement('time');
+      let time = videojs.formatTime(item.duration);
+      duration.className = 'vjs-playlist-duration';
+      duration.setAttribute('datetime', 'PT0H0M' + item.duration + 'S');
+      duration.appendChild(document.createTextNode(time));
+      li.appendChild(duration);
+    }
+
+    // Name and description
+    let name = document.createElement('cite');
+    name.className = 'vjs-playlist-name';
+    name.appendChild(document.createTextNode(item.name || this.localize('Untitled Video')));
+    li.appendChild(name);
+
+    if (item.description) {
+      let description = document.createElement('p');
+      description.className = 'vjs-playlist-description';
+      description.appendChild(document.createTextNode(item.description));
+      li.appendChild(description);
+    }
+    return li;
+  }
+});
+
+videojs.PlaylistMenu = videojs.Component.extend({
+  init: function(player, options) {
+
+    if (!player.playlist || player.playlist() === undefined) {
+      throw new Error('videojs-playlist is required for the playlist component');
+    }
+
+    if (!options.el) {
+      this.el_ = document.createElement('ol');
+      this.el_.className = options.className;
+      options.el = this.el_;
+    }
+
+    videojs.Component.call(this, player, options);
+    this.createPlaylist_();
+
+    if (!videojs.TOUCH_ENABLED) {
+      this.addClass('vjs-mouse');
+    }
+
+    player.on('loadstart', (event) => {
+      this.update();
+    });
+  },
+  createPlaylist_: function() {
+    const playlist = this.player_.playlist();
+
+    this.items = [];
+    for (let i in playlist) {
```

Comment:
I've used `for...in` exclusively to avoid the runtime baggage of `for...of`. `Array#map` also requires shimming. I prefer functional syntax too but it's extra code at runtime just for some syntactic sugar.